app/action/search-actions.coffee

```diff
@@ -96,24 +81,125 @@ getPeliasDataOrEmptyArray = (input, geolocation) ->
 
   return XhrPromise.getJson(config.URL.PELIAS, opts).then (res) -> res.features
 
-directGeocode = (actionContext, input) ->
+
+queryGraphQL = (q, opts) ->
+  payload = JSON.stringify (query: q, variables: null)
+  return XhrPromise.postJson(config.URL.OTP + "index/graphql", opts, payload)
+
+mapRoutes = (res) ->
+  if res
+    res.map (item) ->
+      type: "Route"
+      agency: item.agency
+      properties:
+        label: item.shortName + " " + item.longName
+        layer: 'route-' + item.type
+        link: '/linjat/' + item.patterns[0].code
+      geometry:
+        coordinates:
+          [item.lat, item.lon]
+  else
+    []
+
+mapStops = (res) ->
+  if res
+    res.map (item) ->
+      type: "Stop"
+      properties:
+        label: item.name
+        layer: 'stop'
+        link: '/pysakit/' + item.gtfsId
+      geometry:
+        coordinates:
+          [item.lon, item.lat]
+  else
+    []
+
+getEndpointGTFSResult = (input, reference) ->
+  if input == undefined or input == null or input.trim().length < 3
+    return Promise.resolve []
+
+  queryGraphQL('{' + 'stops(name:"' + input + '") {gtfsId lat lon name}' + '}').then (response) ->
+    mapStops(response?.data?.stops)
+
+
+getCommonGTFSResult = (input, reference, favourites) ->
+  searches = []
+  fav = favourites.map (f) -> '"' + f + '"'
+  searches.push('favouriteRoutes:routes(ids:[' + fav.join(',') + ']) {patterns {code} agency {name} shortName type longName}')
+
+  if input != undefined &&  input != null and input.trim().length > 0
+    doRouteSearch = doStopSearch = false
+
+    #dumb quess on what the user is trying to do
+    isNumber = input.match(/^\d+$/) != null
+    if isNumber
+      lnLen = input.match(/^\d+$/).length
+      if lnLen <= 3
+        doRouteSearch = true
+
+      if lnLen == 4
+        doStopSearch = true
+    else
+      doRouteSearch = doStopSearch = true
+
+  if doRouteSearch then searches.push 'routes(name:"' + input + '") {patterns {code} agency {name} shortName type longName}'
+  if doStopSearch then searches.push 'stops(name:"' + input + '") {gtfsId lat lon name}'
+
+  if searches.length > 0
+    suggestions = []
+    refLatLng = geoUtils.getLatLng(reference.lat, reference.lon)
+    return queryGraphQL('{' + searches.join(' ') + '}').then (response) ->
+
+      favouriteRoutes = mapRoutes(response?.data?.favouriteRoutes).map (favourite) ->
+        favourite.properties.layer = 'favourite'
+        favourite.type= "Favourite"
+        favourite
+      [].concat sortBy(favouriteRoutes, (item) -> ['agency.name', 'properties.label'])
+      .concat sortBy(mapStops(response?.data?.stops), (item) ->
+        stopPos = geoUtils.getLatLng(item.geometry.coordinates[1], item.geometry.coordinates[0])
+        Math.round(stopPos.distanceTo(refLatLng) / 50000))
```

Comment:
Are we using this magical "50000" somewhere else? If so, I would somehow combine this into function that tells what's it doing